File: 백준/Gold/2176. 합리적인 이동경로/합리적인 이동경로.swift
import Foundation

struct Heap<T: Comparable> {
    private var heap: [T?] = [nil]
    var isEmpty: Bool {
        return heap.count <= 1
    }

    mutating func insert(_ element: T) {
        heap.append(element)
        heapifyUp()
    }

    mutating func pop() -> T? {
        guard !isEmpty else { return nil }
        defer { heapifyDown() }

        heap.swapAt(1, heap.count - 1)
        return heap.removeLast()
    }

    private mutating func heapifyUp() {
        var index = heap.count - 1

        while
            index > 1,
            let parent = heap[index / 2],
            let child = heap[index],
            parent > child
        {
            heap.swapAt(index, index / 2)
            index /= 2
        }
    }

    mutating func heapifyDown() {
        var index = 1

        while index * 2 < heap.count {
            var candidate = index * 2

            if
                candidate + 1 < heap.count,
                let left = heap[candidate],
                let right = heap[candidate + 1],
                left > right
            { candidate += 1 }

            guard
                let child = heap[candidate],
                let parent = heap[index],
                parent > child
            else { return }

            heap.swapAt(index, candidate)
            index = candidate
        }
    }
}

struct Edge: Comparable {
    let to: Int
    let weight: Int

    init(_ to: Int, _ weight: Int) {
        self.to = to
        self.weight = weight
    }

    static func <(_ lhs: Edge, _ rhs: Edge) -> Bool {
        return lhs.weight < rhs.weight
    }
}

// MARK: 변수 및 상수
let numbers = readLine()!.split(separator: " ").map { Int(String($0))! }
let nodeCount = numbers[0]
let edgeCount = numbers[1]
var dist = Array(repeating: Int.max, count: nodeCount + 1)
var edges: [[Edge]] = Array(repeating: [], count: nodeCount + 1)
var routes = Array(repeating: -1, count: nodeCount + 1)

// 입력
for _ in 0..<edgeCount {
    let input = readLine()!.split(separator: " ").map { Int(String($0))! }
    let nodeA = input[0]
    let nodeB = input[1]

    edges[nodeA].append(Edge(nodeB, input[2]))
    edges[nodeB].append(Edge(nodeA, input[2]))
}

// MARK: - 풀이
func dijkstra() {
    var heap = Heap<Edge>()
    dist[2] = 0

    heap.insert(Edge(2, 0))

    while let edge = heap.pop() {
        let cur = edge.to
        let cost = edge.weight
        if dist[cur] > cost { continue }
        for edge in edges[cur] {
            let next = edge.to
            let edgeCost = edge.weight

            guard dist[cur] + edgeCost < dist[next] else { continue }

            dist[next] = dist[cur] + edgeCost
            heap.insert(Edge(next, dist[next]))
        }
    }
}

func findRoute(_ node: Int) -> Int {
    if node == 2 { return 1 }
    if routes[node] != -1 { return routes[node] }

    routes[node] = 0
    for edge in edges[node] {
        let next = edge.to
        guard dist[node] > dist[next] else { continue }
        routes[node] += findRoute(next)
    }
    return routes[node]
}

func solution() {
    dijkstra()
    print(findRoute(1))
}

solution()

File: 백준/Gold/2176. 합리적인 이동경로/합리적인 이동경로.cc
#include <iostream>
#include <vector>
#include <algorithm>
#include <queue>
#include <climits>
#define rep(i, a, b) for(auto i = a; i < b; ++i)
#define REP(i, a, b) for(auto i = a; i <= b; ++i)
#define FAST ios_base::sync_with_stdio(false); cin.tie(NULL); cout.tie(NULL)

using namespace std;

// 데이터 구조 
struct Edge {
    int next;
    int cost;

    bool operator<(const Edge& rhs) const {
        return cost > rhs.cost;
    }
};

// 변수 및 상수
int nodeCount, edgeCount;
vector<int> routes;
vector<int> dist;
vector<vector<Edge>> edges;

void dijkstra() {
    priority_queue<Edge> pq;
    pq.push({2, 0});
    dist[2] = 0;

    while (!pq.empty()) {
        Edge edge = pq.top();
        pq.pop();

        int node = edge.next;
        int cost = edge.cost;

        if (dist[node] < cost) { continue; }

        rep(i, 0, edges[node].size()) {
            Edge edge = edges[node][i];
            int next = edge.next;
            int newCost = edge.cost + dist[node];
            
            if (dist[next] <= newCost) { continue; }

            dist[next] = newCost;
            pq.push({next, newCost});
        }
    }
}

int findRoute(int node) {
    if (node == 2) { return 1; }
    if (routes[node] != -1) { return routes[node]; }

    routes[node] = 0;

    rep(i, 0, edges[node].size()) {
        // 합리적인 이동경로인지 확인
        Edge edge = edges[node][i];
        int next = edge.next;

        if (dist[next] >= dist[node]) { continue; }
        routes[node] += findRoute(next);
    }

    return routes[node];
}

int main() {
    FAST;
    cin >> nodeCount >> edgeCount;
    routes.resize(nodeCount + 1, -1);
    dist.resize(nodeCount + 1, INT_MAX);
    edges.resize(nodeCount + 1);

    rep(i, 0, edgeCount) {
        int nodeA, nodeB, cost;
        cin >> nodeA >> nodeB >> cost;

        edges[nodeA].push_back({nodeB, cost});
        edges[nodeB].push_back({nodeA, cost});
    }

    dijkstra(); 
    cout << findRoute(1);
}